File: project-{topic}-advanced/src/main/java/com/alxmobilebackend/controller/ProductController.java
```java
package com.alxmobilebackend.controller;

import com.alxmobilebackend.dto.ProductDto;
import com.alxmobilebackend.service.ProductService;
import com.alxmobilebackend.util.Constants;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping(Constants.PRODUCTS_PATH)
@RequiredArgsConstructor
@SecurityRequirement(name = "bearerAuth")
@Tag(name = "Product Management", description = "API for managing products")
public class ProductController {

    private final ProductService productService;

    @Operation(summary = "Create a new product", description = "Requires ADMIN role")
    @PostMapping
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ProductDto.ProductResponse> createProduct(@Valid @RequestBody ProductDto.ProductCreateRequest request) {
        ProductDto.ProductResponse product = productService.createProduct(request);
        return new ResponseEntity<>(product, HttpStatus.CREATED);
    }

    @Operation(summary = "Get product by ID", description = "Accessible to all authenticated users")
    @GetMapping("/{id}")
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public ResponseEntity<ProductDto.ProductResponse> getProductById(@PathVariable Long id) {
        ProductDto.ProductResponse product = productService.getProductById(id);
        return ResponseEntity.ok(product);
    }

    @Operation(summary = "Get all products with pagination", description = "Accessible to all authenticated users")
    @GetMapping
    @PreAuthorize("hasAnyRole('USER', 'ADMIN')")
    public ResponseEntity<Page<ProductDto.ProductResponse>> getAllProducts(Pageable pageable) {
        Page<ProductDto.ProductResponse> products = productService.getAllProducts(pageable);
        return ResponseEntity.ok(products);
    }

    @Operation(summary = "Update a product by ID", description = "Requires ADMIN role")
    @PutMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    public ResponseEntity<ProductDto.ProductResponse> updateProduct(@PathVariable Long id,
                                                                  @Valid @RequestBody ProductDto.ProductUpdateRequest request) {
        ProductDto.ProductResponse updatedProduct = productService.updateProduct(id, request);
        return ResponseEntity.ok(updatedProduct);
    }

    @Operation(summary = "Delete a product by ID", description = "Requires ADMIN role")
    @DeleteMapping("/{id}")
    @PreAuthorize("hasRole('ADMIN')")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void deleteProduct(@PathVariable Long id) {
        productService.deleteProduct(id);
    }
}
```

File: project-{topic}-enhanced/server/src/modules/products/product.controller.ts
import { Request, Response, NextFunction } from 'express';
import httpStatus from 'http-status';
import * as productService from './product.service';
import { catchAsync } from '@/utils/catchAsync';
import { ApiError } from '@/utils/ApiError';
import logger from '@/utils/logger';

// Create a new product
export const createProduct = catchAsync(async (req: Request, res: Response) => {
  const product = await productService.createProduct(req.body);
  res.status(httpStatus.CREATED).send(product);
});

// Get all products
export const getProducts = catchAsync(async (req: Request, res: Response) => {
  const products = await productService.getProducts();
  res.status(httpStatus.OK).send(products);
});

// Get a single product by ID
export const getProduct = catchAsync(async (req: Request, res: Response) => {
  const product = await productService.getProductById(req.params.productId);
  if (!product) {
    throw new ApiError(httpStatus.NOT_FOUND, 'Product not found');
  }
  res.status(httpStatus.OK).send(product);
});

// Update a product by ID
export const updateProduct = catchAsync(async (req: Request, res: Response) => {
  const product = await productService.updateProductById(req.params.productId, req.body);
  res.status(httpStatus.OK).send(product);
});

// Delete a product by ID
export const deleteProduct = catchAsync(async (req: Request, res: Response) => {
  await productService.deleteProductById(req.params.productId);
  res.status(httpStatus.NO_CONTENT).send();
});